hr_timesheet_sheet_role/models/hr_timesheet_sheet.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2018 Brainbean Apps (https://brainbeanapps.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models, api, _
+from odoo.exceptions import ValidationError
+
+
+class HrTimesheetSheet(models.Model):
+    _inherit = 'hr_timesheet.sheet'
+
+    add_line_role_id = fields.Many2one(
+        comodel_name='project.role',
+        string='Select Role',
+        domain=lambda self: self._domain_add_line_role_id(),
+        help=(
+            'If selected, the associated role is added '
+            'to the timesheet sheet when clicked the button.'
+        ),
+        states={
+            'draft': [('readonly', False)],
+        },
+    )
+    add_line_is_role_required = fields.Boolean(
+        related='add_line_project_id.is_timesheet_role_required',
+    )
+    add_line_limit_role_to_assignments = fields.Boolean(
+        related='add_line_project_id.limit_timesheet_role_to_assignments',
+    )
+
+    @api.multi
+    @api.constrains('company_id', 'add_line_role_id')
+    def _check_company_id_add_line_role_id(self):
+        for sheet in self:
+            if sheet.company_id and sheet.add_line_role_id.company_id and \
+                    sheet.company_id != sheet.add_line_role_id.company_id:
+                raise ValidationError(_(
+                    'The Company in the Timesheet Sheet and in the Role must'
+                    ' be the same.'
+                ))
+
+    @api.onchange('add_line_project_id')
+    def onchange_add_project_id(self):
+        result = super().onchange_add_project_id()
+
+        result['domain'].update({
+            'add_line_role_id': self._domain_add_line_role_id(),
+        })
+
+        return result
+
+    def _domain_add_line_role_id(self):
+        user_id = self.env['res.users'].browse(
```

Comment:
@alexey-pelykh Why not `user_id = self.user_id`? IMO `onchange_add_project_id` should also depend on `user_id` otherwise the domain will not be recomputed if the employee id is changed on the timesheet. `https://github.com/OCA/hr-timesheet/blob/12.0/hr_timesheet_sheet/models/hr_timesheet_sheet.py#L78`

Reply:
Valid point, fixed!